### ISSUE TITLE
Angular Testing

### ISSUE THREAD

**Issue by mattcfilbert** [excellalabs/javascript-project-guidelines#21]
Angular Testing

TLDR of outcome of discussion:
Angular Testing
E2E/Functional: Cypress massively triumphs over Protractor et al
Cypress is framework agnostic and is also our solution for React E2E (feeds two birds with one scone to use it in both places)
Unit & (Cross-Component) Integration
This was a tight vote with Jasmine/Karma narrowly edging out the pairing of Jest and Angular Testing Library
While Jasmine/Karma won the vote, we were running short on time for discussion. Do we want to continue that discussion (and the discussion of overall testing practices, such as diamond vs. pyramid) in the next JSDB session?

---

Unit
-Karma/Jasmine are the defaults from the CLI
- solid setup that folks are comfortable with
in the past, there have been performance issues, but these issues have been addressed
illustrative of an ever improving solution
Angular package benefits (due to it being the Angular solution) - no need to monitor dependencies in the same way you would need to do with an external solution
-Jest
has anyone actually used Jest in Angular?
“it worked fine on a smaller scale”
aligns with React unit testing (though it is bolstered with React Testing Library on the React side - though potential leveraging of Angular Testing Library could further align processes)
potentially faster than Karma/Jasmine
Jest and Karma are VERY similar on the base level (different command names, but effectively do the same thing)
-Mocha
Popular testing library, but potentially at odds with the “Angular Way” of giving you everything out of the box
Is that a problem tho? Works fine
Integration (Cross-Component)
-Angular Testing Library
-Karma/Jasmine (again)
HIGH LEVEL THEORY SECTION** 
Do we have separate files for int tests and unit tests?
diamond structure inherently splits out some testing due to differing priorities, though blending occurs
feature level int tests, reusable component unit tests (+1)
how to sell this? 
unit tests are great with a single non-integrated piece of functionality
BUT we build features to help human being use apps
testing the intended use of features involve components interlocking successfully - a la, int tests
line coverage doesn’t necessarily tell you “do things work together”
Diamond vs pyramid thinking
	-Int tests could be more important for non-reusable components
	-covering a reusable component in unit tests is more crucial (determining it functions correctly in different contexts)
	- ^ diamond (relatively more int tests) vs. pyramid design (base of unit tests)
What is a good FE int test? (e.g. does it need a simulated page environment, or just the involved components?)
E2E / Functional
- Protractor
— Default with the Angular CLI
— Suggested approach from Angular?
— “It’s fine”
— Doesn’t necessarily give the non-tech folk easy access (e.g. gherkin)
— protractor commands can set up mocked data (more naturally hooked in than something like Cypress)
— different browsers can be tested in
-Cypress (WINNER)
— this was our decision for React - could be a boon to have it for both?
— framework agnostic
— typically used in Chrome, but supports Edge, Firefox (potentially others)
— should an E2E test only be done in a way a user would be able to do it? (e.g. not seeding data)
— - Cypress is good for going with this approach, but does not give you a good way to seed test data w/o an API
— you can mock calls to return data (despite not being able to seed DBs without an API / other external middleware
-Puppeteer
— only works in Chrome and Chromium
— potentially more configurability, but certainly more set up
— likely also framework agnostic
